From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Wed, 29 Jan 2025 18:29:19 +0200
Subject: [PATCH] Better EntityShootBowEvent cancellation

Don't increase stats / play bow sound if event was cancelled

diff --git a/net/minecraft/world/item/BowItem.java b/net/minecraft/world/item/BowItem.java
index 7d846ac6163873abee29e860a066ffc130a4ea98..dae742612f539d5335deea4d3ddda2621dbf1fa5 100644
--- a/net/minecraft/world/item/BowItem.java
+++ b/net/minecraft/world/item/BowItem.java
@@ -44,7 +44,7 @@ public class BowItem extends ProjectileWeaponItem {
                 } else {
                     List<ItemStack> list = draw(stack, projectile, player, !projectile.has(net.minecraft.core.component.DataComponents.INTANGIBLE_PROJECTILE)); // Kiterino - Infinite projectiles
                     if (level instanceof ServerLevel serverLevel && !list.isEmpty()) {
-                        this.shoot(serverLevel, player, player.getUsedItemHand(), stack, list, powerForTime * 3.0F, (float) serverLevel.purpurConfig.bowProjectileOffset, powerForTime == 1.0F, null, powerForTime); // Paper - Pass draw strength // Purpur - Projectile offset config
+                        if (!this.shoot(serverLevel, player, player.getUsedItemHand(), stack, list, powerForTime * 3.0F, (float) serverLevel.purpurConfig.bowProjectileOffset, powerForTime == 1.0F, null, powerForTime)); // Paper - Pass draw strength  // Purpur - Projectile offset config // Kiterino - Better EntityShootBowEvent cancellation
                     }
 
                     level.playSound(
diff --git a/net/minecraft/world/item/CrossbowItem.java b/net/minecraft/world/item/CrossbowItem.java
index c4279f25223eacbb708134b4562c0cd042161a30..419ae5c9edd1744992a419ea11c3cf1176388184 100644
--- a/net/minecraft/world/item/CrossbowItem.java
+++ b/net/minecraft/world/item/CrossbowItem.java
@@ -174,7 +174,7 @@ public class CrossbowItem extends ProjectileWeaponItem {
         if (level instanceof ServerLevel serverLevel) {
             ChargedProjectiles chargedProjectiles = weapon.set(DataComponents.CHARGED_PROJECTILES, ChargedProjectiles.EMPTY);
             if (chargedProjectiles != null && !chargedProjectiles.isEmpty()) {
-                this.shoot(serverLevel, shooter, hand, weapon, chargedProjectiles.getItems(), velocity, inaccuracy, shooter instanceof Player, target, 1); // Paper - Pass draw strength
+                if (!this.shoot(serverLevel, shooter, hand, weapon, chargedProjectiles.getItems(), velocity, inaccuracy, shooter instanceof Player, target, 1)); // Paper - Pass draw strength // Kiterino - Better EntityShootBowEvent cancellation
                 if (shooter instanceof ServerPlayer serverPlayer) {
                     CriteriaTriggers.SHOT_CROSSBOW.trigger(serverPlayer, weapon);
                     serverPlayer.awardStat(Stats.ITEM_USED.get(weapon.getItem()));
diff --git a/net/minecraft/world/item/ProjectileWeaponItem.java b/net/minecraft/world/item/ProjectileWeaponItem.java
index aa1cf1653d33cd35f77db28dbee0c2319e1975f3..6b6b02ceb5a152485f39bcb9da7a3b94af67bc1b 100644
--- a/net/minecraft/world/item/ProjectileWeaponItem.java
+++ b/net/minecraft/world/item/ProjectileWeaponItem.java
@@ -40,7 +40,7 @@ public abstract class ProjectileWeaponItem extends Item {
 
     public abstract int getDefaultProjectileRange();
 
-    protected void shoot(
+    protected boolean shoot( // Kiterino - void -> boolean - Better EntityShootBowEvent cancellation
         ServerLevel level,
         LivingEntity shooter,
         InteractionHand hand,
@@ -70,7 +70,7 @@ public abstract class ProjectileWeaponItem extends Item {
                 org.bukkit.event.entity.EntityShootBowEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityShootBowEvent(shooter, weapon, itemStack, projectile, hand, drawStrength, true);
                 if (event.isCancelled()) {
                     event.getProjectile().remove();
-                    return;
+                    return false; // Kiterino - Better EntityShootBowEvent cancellation
                 }
 
                 if (event.getProjectile() == projectile.getBukkitEntity()) {
@@ -79,7 +79,7 @@ public abstract class ProjectileWeaponItem extends Item {
                         level,
                         itemStack
                     ).isRemoved()) {
-                        return;
+                        return false; // Kiterino - Better EntityShootBowEvent cancellation
                     }
                 }
                 // CraftBukkit end
@@ -89,6 +89,7 @@ public abstract class ProjectileWeaponItem extends Item {
                 }
             }
         }
+        return true; // Kiterino - Better EntityShootBowEvent cancellation
     }
 
     protected int getDurabilityUse(ItemStack stack) {
